common/src/jni/main/cpp/NativeCrypto.cpp

```diff
@@ -4311,6 +4311,202 @@ static void NativeCrypto_ASN1_TIME_to_Calendar(JNIEnv* env, jclass, jlong asn1Ti
+    JNI_TRACE("asn1_read_is_empty(%p)", cbs);
+
+    bool empty = (CBS_len(cbs->cbs.get()) == 0);
+    JNI_TRACE("asn1_read_is_empty(%p) => %s", cbs, empty ? "true" : "false");
+    return empty;
+}
+
+static void NativeCrypto_asn1_read_free(CONSCRYPT_UNUSED JNIEnv* env, jclass, jlong cbsRef) {
+    if (cbsRef == 0) {
+        JNI_TRACE("asn1_read_free(0)");
+        return;
+    }
+    CbsHandle* cbs = reinterpret_cast<CbsHandle*>(static_cast<uintptr_t>(cbsRef));
+    JNI_TRACE("asn1_read_free(%p)", cbs);
+    delete cbs;
+}
+
+static jlong NativeCrypto_asn1_write_init(JNIEnv* env, jclass) {
+    JNI_TRACE("asn1_write_init");
+    std::unique_ptr<CBB> cbb(new CBB());
+    if (!CBB_init(cbb.get(), 128)) {
+        Errors::throwIOException(env, "Error writing ASN.1 encoding");
+        return 0;
+    }
+    JNI_TRACE("asn1_write_init => %p", cbb.get());
+    return reinterpret_cast<uintptr_t>(cbb.release());
+}
+
+static jlong NativeCrypto_asn1_write_sequence(JNIEnv* env, jclass, jlong cbbRef) {
+    CBB* cbb = reinterpret_cast<CBB*>(static_cast<uintptr_t>(cbbRef));
+    JNI_TRACE("asn1_write_sequence(%p)", cbb);
+
+    std::unique_ptr<CBB> seq(new CBB());
+    if (!CBB_add_asn1(cbb, seq.get(), CBS_ASN1_SEQUENCE)) {
+        Errors::throwIOException(env, "Error writing ASN.1 encoding");
+        return 0;
+    }
+    JNI_TRACE("asn1_write_sequence(%p) => %p", cbb, seq.get());
+    return reinterpret_cast<uintptr_t>(seq.release());
+}
+
+static void NativeCrypto_asn1_write_octetstring(JNIEnv* env, jclass, jlong cbbRef, jbyteArray data) {
+    CBB* cbb = reinterpret_cast<CBB*>(static_cast<uintptr_t>(cbbRef));
+    JNI_TRACE("asn1_write_octetstring(%p, %p)", cbb, data);
+
+    ScopedByteArrayRO bytes(env, data);
+    if (bytes.get() == nullptr) {
+        JNI_TRACE("asn1_write_octetstring(%p, %p) => using byte array failed", cbb, data);
+        return;
+    }
+
+    std::unique_ptr<CBB> octetstring(new CBB());
+    if (!CBB_add_asn1(cbb, octetstring.get(), CBS_ASN1_OCTETSTRING)) {
+        Errors::throwIOException(env, "Error writing ASN.1 encoding");
+        return;
+    }
+    if (!CBB_add_bytes(octetstring.get(), reinterpret_cast<const uint8_t*>(bytes.get()), bytes.size())) {
+        Errors::throwIOException(env, "Error writing ASN.1 encoding");
+        return;
+    }
+}
+
+static void NativeCrypto_asn1_write_integer(JNIEnv* env, jclass, jlong cbbRef, jlong data) {
+    CBB* cbb = reinterpret_cast<CBB*>(static_cast<uintptr_t>(cbbRef));
+    JNI_TRACE("asn1_write_integer(%p)", cbb);
+
+    if (!CBB_add_asn1_uint64(cbb, static_cast<uint64_t>(data))) {
+        Errors::throwIOException(env, "Error writing ASN.1 encoding");
+        return;
+    }
+}
+
+static jbyteArray NativeCrypto_asn1_write_finish(JNIEnv* env, jclass, jlong cbbRef) {
+    CBB* cbb = reinterpret_cast<CBB*>(static_cast<uintptr_t>(cbbRef));
+    JNI_TRACE("asn1_write_finish(%p)", cbb);
+
+    uint8_t* data;
```

Comment:
This needs to be unique_ptr-ified if CBB_finish succeeds so that we call OPENSSL_free even if the stuff below fails.